### PR TITLE
Add CoreCLR personality function and EH table emission

### PR DESCRIPTION
**Pull request by JosephTremoulet**
This is a PR for changes going into the newly-created EH branch.  The purpose is to park the changes-so-far based on the soon-to-be-abandoned WinEH approach of using Itanium-style IR paired with funclet outlining in WinEHPrepare.  Support will be reworked when [the new IR representation](http://thread.gmane.org/gmane.comp.compilers.llvm.devel/85783) lands.  I'm going to shift focus to aspects of EH that are independent of the impending rewrite (e.g. filter function outlining).  Putting these changes here lets anyone curious see where things are/were heading and provide feedback, and lets me park similar changes dependent on this in the EH branch of LLILC.

### CODE REVIEW

**Comment by JosephTremoulet** [comment]
Something along the lines of `emitCLRExceptionTable` (which emits the extra xdata after unwind infos consumed by dotnet/llilc#606) is likely to stick, as is adding the CLR entry in `classifyEHPersonality`.  The rest (most of which reconstructs the handler tree from the `eh.actions` left behind by outlining) will likely be redone when the IR is changed and the outliner removed.

File: lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp [hunk 1]
```diff
@@ -318,6 +320,60 @@ void FunctionLoweringInfo::addSEHHandlersForLPads(
...
+      if (Current == nullptr) {
+        Current = new HandlerTreeNode();
+        Current->Parent = Parent;
+        Current->Handler = Handler;
+        CatchHandler *TheCatchHandler = dyn_cast<CatchHandler>(Action);
+        if (TheCatchHandler == nullptr) {
+          assert(isa<CleanupHandler>(Action));
+          Current->CatchType = 0;
+        } else {
+          assert(isa<Operator>(TheCatchHandler->getSelector()) && "Expected IntToPtr(handkerToken)");
```
Comment (AndyAyersMS) [inline lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp]: `handker` --> `handler`

File: lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp [hunk 2]
```diff
@@ -318,6 +320,60 @@ void FunctionLoweringInfo::addSEHHandlersForLPads(
...
+          Operator *IntToPtr = cast<Operator>(TheCatchHandler->getSelector());
+          assert(IntToPtr->getOpcode() == CastInst::IntToPtr && "Expected IntToPtr(handkerToken)");
```
Comment (AndyAyersMS) [inline lib/CodeGen/SelectionDAG/FunctionLoweringInfo.cpp]: ditto

File: lib/MC/MCWin64EH.cpp [hunk 3]
```diff
@@ -206,10 +207,16 @@ static void EmitUnwindInfo(MCStreamer &streamer, WinEH::FrameInfo *info) {
   if (flags & (Win64EH::UNW_ChainInfo << 3))
     EmitRuntimeFunction(streamer, info->ChainedParent);
   else if (flags &
-           ((Win64EH::UNW_TerminateHandler|Win64EH::UNW_ExceptionHandler) << 3))
-    streamer.EmitValue(MCSymbolRefExpr::create(info->ExceptionHandler,
-                                              MCSymbolRefExpr::VK_COFF_IMGREL32,
-                                              context), 4);
+           ((Win64EH::UNW_TerminateHandler|Win64EH::UNW_ExceptionHandler) << 3)) {
+    if (classifyEHPersonality(info->ExceptionHandler->getName()) == EHPersonality::CoreCLR) {
```
Comment (AndyAyersMS) [inline lib/MC/MCWin64EH.cpp]: You could instead check if the runtime environment is CoreCLR (and assert if you see any unexpected personality routine if so).

**Comment by AndyAyersMS** [comment]
LGTM, just a few small notes.

**Comment by JosephTremoulet** [comment]
Updated as suggested; merging into EH branch.

**Pull request merged**